### ISSUE TITLE
Discriminator loss

### ISSUE THREAD

**Issue by amans191** [PacktPublishing/Generative-Adversarial-Networks-Projects#4]
Hi, I have been training my model, and it doesn't seem to be changing from a cube to a chair. I trained it for 10,000 epochs and it was still a cube. How many epochs did you train it for before the graph started to converge and your result looked like a chair?

Also, my discriminator loss graph keeps going up and down, whereas your one in the book seems to be steadily decreasing, why is this happening. Please see my discriminator loss graph below. 

<img width="947" alt="Screenshot 2019-06-26 at 11 34 11" src="https://user-images.githubusercontent.com/15716505/60205767-f90e2880-9849-11e9-8017-24bcd331c820.png">

**Comment by kailashahirwar** [comment]
Hey @amans191 Can you share your training part with me?

**Comment by amans191** [comment]
`if __name__ == '__main__':
    """
    Specify Hyperparameters
    """
    object_name = "chair"
    data_dir = "data/3DShapeNets/volumetric_data/" \
               "{}/30/train/*.mat".format(object_name)
    gen_learning_rate = 0.0025
    dis_learning_rate = 10e-5
    beta = 0.5
    batch_size = 1
    z_size = 200
    epochs = 10000
    MODE = "train"

    """
    Create models
    """
    gen_optimizer = Adam(lr=gen_learning_rate, beta_1=beta)
    dis_optimizer = Adam(lr=dis_learning_rate, beta_1=beta)

    discriminator = build_discriminator()
    discriminator.compile(loss='binary_crossentropy', optimizer=dis_optimizer)

    generator = build_generator()
    generator.compile(loss='binary_crossentropy', optimizer=gen_optimizer)

    discriminator.trainable = False

    input_layer = Input(shape=(1, 1, 1, z_size))
    generated_volumes = generator(input_layer)
    validity = discriminator(generated_volumes)
    adversarial_model = Model(inputs=[input_layer], outputs=[validity])
    adversarial_model.compile(loss='binary_crossentropy', optimizer=gen_optimizer)

    print("Loading data...")
    volumes = get3DImages(data_dir=data_dir)
    volumes = volumes[..., np.newaxis].astype(np.float)
    print("Data loaded...")

    tensorboard = TensorBoard(log_dir="logs/{}".format(time.time()))
    tensorboard.set_model(generator)
    tensorboard.set_model(discriminator)

    labels_real = np.reshape(np.ones((batch_size,)), (-1, 1, 1, 1, 1))
    labels_fake = np.reshape(np.zeros((batch_size,)), (-1, 1, 1, 1, 1))

    if MODE == 'train':
        for epoch in range(epochs):
            print("Epoch:", epoch)

            gen_losses = []
            dis_losses = []

            number_of_batches = int(volumes.shape[0] / batch_size)
            print("Number of batches:", number_of_batches)
            for index in range(number_of_batches):
                print("Batch:", index + 1)

                z_sample = np.random.normal(0, 0.33, size=[batch_size, 1, 1, 1, z_size]).astype(np.float32)
                volumes_batch = volumes[index * batch_size:(index + 1) * batch_size, :, :, :]

                # Next, generate volumes using the generate network
                gen_volumes = generator.predict_on_batch(z_sample)

                """
                Train the discriminator network
                """
                discriminator.trainable = True
                if index % 2 == 0:
                    loss_real = discriminator.train_on_batch(volumes_batch, labels_real)
                    loss_fake = discriminator.train_on_batch(gen_volumes, labels_fake)

                    d_loss = 0.5 * np.add(loss_real, loss_fake)
                    print("d_loss:{}".format(d_loss))

                else:
                    d_loss = 0.0

                discriminator.trainable = False
                """
                Train the generator network
                """
                z = np.random.normal(0, 0.33, size=[batch_size, 1, 1, 1, z_size]).astype(np.float32)
                g_loss = adversarial_model.train_on_batch(z, labels_real)
                print("g_loss:{}".format(g_loss))

                gen_losses.append(g_loss)
                dis_losses.append(d_loss)

                # Every 10th mini-batch, generate volumes and save them
                if index % 10 == 0:
                    z_sample2 = np.random.normal(0, 0.33, size=[batch_size, 1, 1, 1, z_size]).astype(np.float32)
                    generated_volumes = generator.predict(z_sample2, verbose=3)
                    for i, generated_volume in enumerate(generated_volumes[:5]):
                        voxels = np.squeeze(generated_volume)
                        voxels[voxels < 0.5] = 0.
                        voxels[voxels >= 0.5] = 1.
                        saveFromVoxels(voxels, "results/img_{}_{}_{}".format(epoch, index, i))

            # Write losses to Tensorboard
            write_log(tensorboard, 'g_loss', np.mean(gen_losses), epoch)
            write_log(tensorboard, 'd_loss', np.mean(dis_losses), epoch)

        """
        Save models
        """
        generator.save_weights(os.path.join("models", "generator_weights.h5"))
        discriminator.save_weights(os.path.join("models", "discriminator_weights.h5"))

    if MODE == 'predict':
        # Create models
        generator = build_generator()
        discriminator = build_discriminator()

        # Load model weights
        generator.load_weights(os.path.join("models", "generator_weights.h5"), True)
        discriminator.load_weights(os.path.join("models", "discriminator_weights.h5"), True)

        # Generate 3D models
        z_sample = np.random.normal(0, 1, size=[batch_size, 1, 1, 1, z_size]).astype(np.float32)
        generated_volumes = generator.predict(z_sample, verbose=3)

        for i, generated_volume in enumerate(generated_volumes[:2]):
            voxels = np.squeeze(generated_volume)
            voxels[voxels < 0.5] = 0.
            voxels[voxels >= 0.5] = 1.
            saveFromVoxels(voxels, "results/gen_{}".format(i))
`

**Comment by enniobianco** [comment]
I also have exactly the same problem.
I'm also producing and storing only red cubes.
Hyperparameters:
gen_learning_rate = 0.0025
dis_learning_rate = 0.00001
beta = 0.5
batch_size = 1
z_size = 200
epochs = 10000

**Comment by kailashahirwar** [comment]
@enniobianco @amans191 I will be reproducing the results and will let you guys know.

**Comment by enniobianco** [comment]
thanks
    Il lunedì 22 luglio 2019, 08:44:24 CEST, Kailash Ahirwar <notifications@github.com> ha scritto:  
 
 
@enniobianco @amans191 I will be reproducing the results and will let you guys know.

**Comment by amans191** [comment]
Could you share your trained models once you run it. The generator_weights.h5 and the discriminator_weights.h5 files

**Comment by enniobianco** [comment]
The files produced must be sent via we transfer because they are very large.But, I can give you access to my computer through AnyDesk. Ok?


    Il lunedì 22 luglio 2019, 10:19:05 CEST, mandersmate <notifications@github.com> ha scritto:  
 
 
Could you share your trained models once you run it. The generator_weights.h5 and the discriminator_weights.h5 files

**Comment by enniobianco** [comment]
https://drive.google.com/file/d/1BM_odp4DzW7nP_JQsL7S4_mci5QZuPqb/view?usp=sharing
https://drive.google.com/file/d/17CbVS5dMzRqLMNgslWzuWihHmY5FOLcO/view?usp=sharing
But did you solve the problem?
Ennio

**Comment by amans191** [comment]
Sorry, I meant the above question for @kailashahirwar, to send his files once he has trained his model.

**Comment by amans191** [comment]
Also, @enniobianco, were you able to change the Batch size, it says in the book to try it with different batch sizes, but every time I change it, the code doesn't run.

**Comment by enniobianco** [comment]
@amans191 yes, it is exactly as you say.
with batch_size = 32 it does not work
works with batch_size = 2, but the result is always a red cube
I'm trying with the directori bed, but ... always a red cube.
all this is unnerving.

**Comment by amans191** [comment]
@kailashahirwar have you reproduced your results yet? and if so could I get the .h5 file?

**Comment by enniobianco** [comment]
@kailashahirwar @amans191 
I observed that 
def plotAndSaveVoxel(file_path, voxel):
    """
    Plot a voxel
    """
    fig = plt.figure()
    ax = fig.gca(projection='3d')
    ax.set_aspect('equal')
    ax.voxels(voxel, edgecolor="red")
    plt.show()
    plt.savefig(file_path)

is never called.
So we see the 3D container but not the voxel image

**Comment by amans191** [comment]
@enniobianco were you able to get it working using that function?

**Comment by enniobianco** [comment]
@amans191 NO. I couldn't get it to work. I simply asked myself why that function that seems to describe the contours of the 3D object was never called.

**Comment by kailashahirwar** [comment]
Hey @amans191 I will have a look at it again.

**Comment by ManikandanKurup-Packt** [comment]
Hi @amans191 @kailashahirwar 
Can we close this if the issue has been resolved?

**Comment by kailashmatelabs** [comment]
@enniobianco @amans191 I did not use "plotAndSaveVoxel", but "saveFromVoxels", which is used to store voxels. If you have found any problem with "saveFromVoxels", let me know.

**Comment by amans191** [comment]
oh pity, only if you were a year early!
